mutable_tree.go

```diff
@@ -848,55 +782,69 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 	if tree.VersionExists(version) {
 		// If the version already exists, return an error as we're attempting to overwrite.
 		// However, the same hash means idempotent (i.e. no-op).
-		existingHash, err := tree.ndb.getRoot(version)
+		existingNodeKey, err := tree.ndb.GetRoot(version)
 		if err != nil {
 			return nil, version, err
 		}
-
-		// If the existing root hash is empty (because the tree is empty), then we need to
-		// compare with the hash of an empty input which is what `WorkingHash()` returns.
-		if len(existingHash) == 0 {
-			existingHash = sha256.New().Sum(nil)
+		var existingRoot *Node
+		if existingNodeKey.version != 0 {
+			existingRoot, err = tree.ndb.GetNode(existingNodeKey)
+			if err != nil {
+				return nil, version, err
+			}
 		}
 
 		newHash, err := tree.WorkingHash()
 		if err != nil {
 			return nil, version, err
 		}
 
-		if bytes.Equal(existingHash, newHash) {
+		if (existingRoot == nil && tree.root == nil) || (existingRoot != nil && bytes.Equal(existingRoot.hash, newHash)) { // TODO with WorkingHash
 			tree.version = version
+			tree.root = existingRoot
 			tree.ImmutableTree = tree.ImmutableTree.clone()
 			tree.lastSaved = tree.ImmutableTree.clone()
-			tree.orphans = map[string]int64{}
-			return existingHash, version, nil
+			tree.orphans = []*NodeKey{}
+			return newHash, version, nil
 		}
 
-		return nil, version, fmt.Errorf("version %d was already saved to different hash %X (existing hash %X)", version, newHash, existingHash)
+		return nil, version, fmt.Errorf("version %d was already saved to different hash from %X (existing nodeKey %d)", version, newHash, existingNodeKey)
 	}
 
-	if tree.root == nil {
-		// There can still be orphans, for example if the root is the node being
-		// removed.
-		logger.Debug("SAVE EMPTY TREE %v\n", version)
-		if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
-			return nil, 0, err
-		}
-		if err := tree.ndb.SaveEmptyRoot(version); err != nil {
-			return nil, 0, err
-		}
-	} else {
-		logger.Debug("SAVE TREE %v\n", version)
-		if _, _, err := tree.ndb.SaveBranch(tree.root); err != nil {
+	logger.Debug("SAVE TREE %v\n", version)
+	// save orphans
+	// orphans := tree.getOrphans()
+	// orphans = append(orphans, tree.orphans...) // should add updated leaves
+	if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
+		return nil, 0, err
+	}
+	// save new nodes
+	if tree.root != nil {
+		// ensure init the nonce
+		tree.nonce = 0
```

Comment:
It seems `tree.nonce` could be a local var?

Reply:
yeah, it sounds good!